File: Practica laboratorio/POO.cpp
#include <iostream>
#include <string>

using namespace std;

// 1. CLASE BASE
class PuntoAcopio {
// 'protected' permite que la clase hija acceda a estas variables sin complicarse
protected: 
    int codigo;
    string barrio;
    double totalRecogido;

public:
    // Constructor
    PuntoAcopio(int c, string b, double t) {
        codigo = c;
        barrio = b;
        totalRecogido = t;
    }

    // Método 1: Sumar kilos
    void registrarRecoleccion(double kilos) {
        totalRecogido = totalRecogido + kilos;
    }

    // Método 2: Revisar meta
    bool superaMeta(double meta) {
        return totalRecogido >= meta;
    }

    // Imprimir datos ('virtual' permite que se sobrescriba)
    virtual void mostrarDescripcion() {
        cout << "Punto " << codigo << " (" << barrio << ") - Total: " << totalRecogido << " kg" << endl;
    }
};

// 2. CLASE HIJA (Hereda de PuntoAcopio)
class MaterialEspecial : public PuntoAcopio {
private:
    string tipo; // Atributo propio

public:
    // Constructor
    MaterialEspecial(int c, string b, double t, string tp) : PuntoAcopio(c, b, t) {
        tipo = tp;
    }

    // Redefinimos el método de la clase padre
    void mostrarDescripcion() override {
        cout << "Punto Especial " << codigo << " (" << barrio << ") - Total: " << totalRecogido << " kg | Tipo: " << tipo << endl;
    }
};

int main() {
    // Arreglo de 4 punteros a la clase base
    PuntoAcopio* centro[4];

    // Creamos los objetos (2 normales y 2 especiales)
    centro[0] = new PuntoAcopio(101, "Centro", 120.5);
    centro[1] = new MaterialEspecial(201, "Norte", 85.0, "Pilas");
    centro[2] = new PuntoAcopio(102, "Sur", 310.0);
    centro[3] = new MaterialEspecial(202, "Occidente", 45.0, "Aceite");

    // Probamos sumar kilos
    centro[0]->registrarRecoleccion(30.0);

    cout << "=== LISTA DE PUNTOS ===" << endl;

    // Recorremos el arreglo
    for (int i = 0; i < 4; i++) {
        centro[i]->mostrarDescripcion(); // Muestra la descripción correspondiente
        
        if (centro[i]->superaMeta(100.0)) {
            cout << "   -> ¡Supero la meta de 100 kg!" << endl;
        } else {
            cout << "   -> No alcanzada la meta." << endl;
        }
    }

    return 0;
}

File: Practica laboratorio/POO.py
class PuntoAcopio:
    def __init__(self, codigo, barrio, total_recogido=0.0):
        # Atributos encapsulados (privados por convención con un guion bajo)
        self._codigo = codigo
        self._barrio = barrio
        self._total_recogido = total_recogido

    # Método 1: Registrar recolección
    def registrar_recoleccion(self, kilos):
        if kilos > 0:
            self._total_recogido += kilos

    # Método 2: Verificar si supera la meta
    def supera_meta(self, meta):
        return self._total_recogido >= meta

    # Getters
    def get_codigo(self):
        return self._codigo

    def get_barrio(self):
        return self._barrio

    def get_total_recogido(self):
        return self._total_recogido

    # Descripción base
    def mostrar_descripcion(self):
        print(f"[PUNTO GENERAL] Código: {self._codigo} | Barrio: {self._barrio} | Total: {self._total_recogido} kg")


# 2. CLASE DERIVADA: MaterialEspecial
class MaterialEspecial(PuntoAcopio):
    def __init__(self, codigo, barrio, total_recogido, tipo_material):
        # Llamada al constructor de la clase base
        super().__init__(codigo, barrio, total_recogido)
        self._tipo_material = tipo_material  # Atributo propio

    # Redefinición del método de la clase padre
    def mostrar_descripcion(self):
        print(f"[MATERIAL ESPECIAL] Código: {self._codigo} | Barrio: {self._barrio} | Total: {self._total_recogido} kg | Tipo: {self._tipo_material}")


# 3. PRUEBA
# Lista mezclando objetos de ambas clases
centro = [
    PuntoAcopio(101, "Centro", 120.5),
    MaterialEspecial(201, "Norte", 85.0, "Baterías y Pilas"),
    PuntoAcopio(102, "Sur", 310.0),
    MaterialEspecial(202, "Occidente", 45.0, "Aceite Usado")
]

# Ejecutar métodos
centro[0].registrar_recoleccion(30.0)
centro[1].registrar_recoleccion(20.0)

print("=== INFORME DE PUNTOS DE ACOPIO ===\n")

# Recorrido de la lista y llamado al método correspondiente (Polimorfismo en Python)
for punto in centro:
    punto.mostrar_descripcion()
    
    if punto.supera_meta(100.0):
        print("   -> Estado: Superó la meta de 100 kg")
    else:
        print("   -> Estado: No ha alcanzado la meta de 100 kg")
    print("-" * 50)